File: impl/python/tests/test_streaming.py
# Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
"""C9 native streaming for the Python SDK, graded against the shared independent corpus
vectors/stream/cases.json (NOT produced by this code): the rolling-SHA-384 commitment over chunks in
absolute-offset order, the mid-stream checkpoint digests, and the StreamOpen / StreamCommit /
StreamCheckpoint body bytes. A native stream is non-repudiable with ONE end-commitment signature, not
N per-chunk signatures; altering any delivered byte invalidates the commitment (StreamDigestMismatch),
and a stream whose effect exceeds the granted ceiling is refused at open, BEFORE any chunk (R-10.3).

NOTE ON THE ORACLE DIR NAME: the Go module is impl/go/streaming/ but its oracle lives at
vectors/stream/cases.json (the directory is `stream`, not `streaming`). This test grades against
vectors/stream/cases.json.

The StreamOpen/StreamCommit/StreamCheckpoint SIGNATURES are real deterministic ML-DSA-65 (a raw
signature over the body, matching impl/go/streaming's s.Sign) demonstrated in isolation; the corpus
carries no signed vector, so signing is NOT corpus-graded (stated honestly, D5).

Written test-first; the streaming module is absent until ported, so this fails RED on import until
impl/python/naalp/streaming.py lands. Mutation: inverting the digest comparison in verify_commit
(!= -> ==) makes a valid stream fail its commitment check and flips test_tamper_invalidates_commit.

Run:  python -m unittest tests.test_streaming      (from impl/python/)
"""
import json
import os
import unittest

from naalp import cose, policy, streaming


def _vectors():
    d = os.path.dirname(os.path.abspath(__file__))
    for _ in range(6):
        p = os.path.join(d, "vectors", "stream", "cases.json")
        if os.path.isfile(p):
            with open(p, encoding="utf-8") as f:
                return json.load(f)
        d = os.path.dirname(d)
    raise FileNotFoundError("vectors/stream/cases.json not found")


ALG = cose.ALG_MLDSA65
SEED = bytes([60]) * 32
PK = cose.mldsa_keygen("ML-DSA-65", SEED)


def hb(s):
    return bytes.fromhex(s)


class StreamingConformance(unittest.TestCase):
    C = _vectors()

    def _chunks(self):
        return [streaming.Chunk(c["offset"], hb(c["data_hex"])) for c in self.C["chunks"]]

    def test_digest_and_bodies_match_oracle(self):
        chunks = self._chunks()
        self.assertEqual(streaming.commit_digest(chunks).hex(), self.C["final_digest_hex"])
        # Rolling: DigestSoFar after each chunk equals the checkpoint, and the last equals final.
        sd = streaming.StreamDigest()
        for i, ch in enumerate(chunks):
            sd.update(ch.data)
            if i < len(self.C["checkpoints"]):
                self.assertEqual(sd.digest_so_far().hex(), self.C["checkpoints"][i]["digest_so_far_hex"],
                                 "checkpoint %d" % i)
        self.assertEqual(sd.digest_so_far().hex(), self.C["final_digest_hex"])

        stream_id = hb(self.C["stream_id_hex"])
        open_ = streaming.StreamOpen(stream_id, self.C["effect"], hb(self.C["approval_hex"]), self.C["substream"])
        self.assertEqual(open_.bytes().hex(), self.C["open_body_hex"])
        commit = streaming.StreamCommit(stream_id, hb(self.C["final_digest_hex"]))
        self.assertEqual(commit.bytes().hex(), self.C["commit_body_hex"])
        cp = streaming.StreamCheckpoint(stream_id, self.C["checkpoints"][0]["through_offset"],
                                        hb(self.C["checkpoints"][0]["digest_so_far_hex"]))
        self.assertEqual(cp.bytes().hex(), self.C["checkpoint_body_hex"])

    def test_tamper_invalidates_commit(self):
        # MUTATION TARGET: the valid stream verifies; altering one delivered byte invalidates the
        # commitment (StreamDigestMismatch) -- R-10.2.
        stream_id = hb(self.C["stream_id_hex"])
        chunks = self._chunks()
        commit = streaming.StreamCommit(stream_id, hb(self.C["final_digest_hex"]))
        self.assertIsNone(streaming.verify_commit(commit, chunks))  # valid stream verifies

        tampered = self._chunks()
        tampered[self.C["tamper"]["chunk_index"]].data = hb(self.C["tamper"]["flipped_data_hex"])
        with self.assertRaises(streaming.StreamError) as cm:
            streaming.verify_commit(commit, tampered)
        self.assertEqual(cm.exception.kind, "StreamDigestMismatch")
        # the tampered rolling digest is exactly the oracle's tamper digest
        self.assertEqual(streaming.commit_digest(tampered).hex(), self.C["tamper"]["digest_hex"])

    def test_checkpoint_verifies_prefix(self):
        stream_id = hb(self.C["stream_id_hex"])
        allc = self._chunks()
        for i, cpj in enumerate(self.C["checkpoints"]):
            cp = streaming.StreamCheckpoint(stream_id, cpj["through_offset"], hb(cpj["digest_so_far_hex"]))
            prefix = allc[:i + 1]  # chunks through this checkpoint, WITHOUT the end
            self.assertIsNone(streaming.verify_checkpoint(cp, prefix), "checkpoint %d" % i)
        # A checkpoint claiming the first prefix but given the full stream fails (wrong length).
        cp0 = streaming.StreamCheckpoint(stream_id, self.C["checkpoints"][0]["through_offset"],
                                         hb(self.C["checkpoints"][0]["digest_so_far_hex"]))
        with self.assertRaises(streaming.StreamError) as cm:
            streaming.verify_checkpoint(cp0, allc)
        self.assertEqual(cm.exception.kind, "StreamDigestMismatch")

    def test_effect_refused_before_chunk(self):
        stream_id = hb(self.C["stream_id_hex"])
        # A destructive stream under a read-only grant is refused (R-10.3).
        destructive = streaming.StreamOpen(stream_id, policy.DESTRUCTIVE, None, 1)
        with self.assertRaises(streaming.StreamError) as cm:
            streaming.open_stream(destructive, policy.READ_ONLY)
        self.assertEqual(cm.exception.kind, "EffectNotAuthorized")
        # The corpus stream (idempotent_write) under an idempotent_write grant is authorized.
        ok = streaming.StreamOpen(stream_id, self.C["effect"], hb(self.C["approval_hex"]), self.C["substream"])
        self.assertIsNone(streaming.open_stream(ok, policy.IDEMPOTENT_WRITE))
        # An unrecognized effect fails closed to destructive and is refused below destructive.
        unknown = streaming.StreamOpen(stream_id, 99, None, 1)
        with self.assertRaises(streaming.StreamError):
            streaming.open_stream(unknown, policy.NON_IDEMPOTENT_WRITE)

    def test_offset_order_matters(self):
        chunks = self._chunks()
        reversed_ = list(reversed(chunks))
        # CommitDigest sorts by absolute offset, so input order is irrelevant.
        self.assertEqual(streaming.commit_digest(chunks).hex(), streaming.commit_digest(reversed_).hex())
        # Swap the data at the first two offsets: same bytes, different positions -> different digest.
        swapped = self._chunks()
        swapped[0].data, swapped[1].data = swapped[1].data, swapped[0].data
        self.assertNotEqual(streaming.commit_digest(swapped).hex(), self.C["final_digest_hex"])

    def test_signed_commitment_in_isolation(self):
        # NOT corpus-graded (no signature vector). Real deterministic ML-DSA raw round-trip: one
        # end-commitment signature covers the whole stream (R-10.2).
        commit = streaming.StreamCommit(hb(self.C["stream_id_hex"]), hb(self.C["final_digest_hex"]))
        sig = streaming.sign_commit(commit, ALG, SEED)
        self.assertTrue(streaming.verify_commit_sig(commit, ALG, PK, sig))
        bad = bytearray(sig)
        bad[0] ^= 0x01
        self.assertFalse(streaming.verify_commit_sig(commit, ALG, PK, bytes(bad)))
        # StreamOpen is also a signed object.
        open_ = streaming.StreamOpen(hb(self.C["stream_id_hex"]), self.C["effect"],
                                     hb(self.C["approval_hex"]), self.C["substream"])
        osig = streaming.sign_open(open_, ALG, SEED)
        self.assertTrue(streaming.verify_open_sig(open_, ALG, PK, osig))


if __name__ == "__main__":
    unittest.main()

File: impl/ruby/test/test_streaming.rb
# Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
#
# C9 native-streaming conformance for the Ruby SDK (design.md §10; R-10.1..10.6), graded against the
# shared independent corpus vectors/stream/cases.json (NOT produced by this code). NOTE the name
# mismatch: the module is impl/ruby/lib/naalp/streaming.rb but the oracle directory is `stream` (not
# `streaming`), matching the Go module streaming/ graded against vectors/stream/. The tests grade the
# deterministic StreamOpen / StreamCommit / StreamCheckpoint body bytes byte-for-byte, the rolling
# SHA-384 commitment over chunks in absolute-offset order, the checkpoint digest_so_far, and the
# fail-closed effect authorization before any chunk.
#
# The LOAD-BEARING C9 property graded here: the whole stream is non-repudiable with ONE commitment
# digest, and altering ANY delivered byte invalidates the commitment (StreamDigestMismatch). The
# StreamOpen/StreamCommit/StreamCheckpoint signatures are real deterministic ML-DSA-65 (a raw signature
# over the body) demonstrated in isolation, but the corpus carries no signature vector, so they are NOT
# corpus-graded (stated honestly; skip-loud where deterministic ML-DSA is unavailable).
#
# Written test-first; the module is absent until ported, so this fails RED (uninitialized constant
# Naalp::Streaming) until impl/ruby/lib/naalp/streaming.rb lands, and a mutation forcing the rolling
# commitment to ignore the chunk data flips test_commit_digest_matches_oracle on its assertion.
#
# Run:  ruby -Ilib -Itest test/test_streaming.rb      (from impl/ruby/)
require 'minitest/autorun'
require 'json'
require 'naalp'

def stream_vectors
  d = File.expand_path(File.dirname(__FILE__))
  6.times do
    p = File.join(d, "vectors", "stream", "cases.json")
    return JSON.parse(File.read(p, encoding: "utf-8")) if File.file?(p)
    d = File.dirname(d)
  end
  raise "vectors/stream/cases.json not found"
end

def hb(hex)
  [hex].pack("H*")
end

ALG = Naalp::COSE::ALG_MLDSA65
SEED = ("\x00" * 32).b

class StreamingConformance < Minitest::Test
  C = stream_vectors

  def stream_id
    hb(C["stream_id_hex"])
  end

  def corpus_chunks
    C["chunks"].map { |c| Naalp::Streaming::Chunk.new(c["offset"], hb(c["data_hex"])) }
  end

  def test_stream_channel_constant
    assert_equal 0x000C, Naalp::Streaming::STREAM_CHANNEL
  end

  # Ruby encoding == the non-circular oracle, byte-for-byte, for the three signed stream objects.
  def test_open_commit_checkpoint_bodies_match_oracle
    open = Naalp::Streaming::StreamOpen.new(stream_id, C["effect"], hb(C["approval_hex"]), C["substream"])
    assert_equal C["open_body_hex"], open.bytes.unpack1("H*"), "StreamOpen body"

    commit = Naalp::Streaming::StreamCommit.new(stream_id, hb(C["final_digest_hex"]))
    assert_equal C["commit_body_hex"], commit.bytes.unpack1("H*"), "StreamCommit body"

    cp0 = C["checkpoints"][0]
    cp = Naalp::Streaming::StreamCheckpoint.new(stream_id, cp0["through_offset"], hb(cp0["digest_so_far_hex"]))
    assert_equal C["checkpoint_body_hex"], cp.bytes.unpack1("H*"), "StreamCheckpoint body"
  end

  # THE MUTATION TARGET. The load-bearing C9 property: the rolling SHA-384 over the chunks in
  # absolute-offset order is the ONE commitment covering the whole stream. Forcing the accumulator to
  # ignore the chunk data collapses every stream to SHA-384("") and flips this on its assertion.
  def test_commit_digest_matches_oracle
    assert_equal C["final_digest_hex"], Naalp::Streaming.commit_digest(corpus_chunks).unpack1("H*"),
                 "rolling commitment over all chunks"
    # every checkpoint's digest_so_far == the rolling digest of the prefix through that offset.
    C["checkpoints"].each do |cpv|
      prefix = corpus_chunks.select { |c| c.offset < cpv["through_offset"] }
      assert_equal cpv["digest_so_far_hex"], Naalp::Streaming.commit_digest(prefix).unpack1("H*"),
                   "checkpoint digest_so_far through #{cpv['through_offset']}"
    end
  end

  # verify_commit accepts the delivered chunks; a tampered byte at a mid-stream offset invalidates the
  # commitment (StreamDigestMismatch). The tampered stream's own digest == the corpus tamper digest.
  def test_tamper_invalidates_commitment
    commit = Naalp::Streaming::StreamCommit.new(stream_id, hb(C["final_digest_hex"]))
    assert_nil Naalp::Streaming.verify_commit(commit, corpus_chunks), "clean delivery verifies"

    t = C["tamper"]
    tampered = corpus_chunks
    tampered[t["chunk_index"]] = Naalp::Streaming::Chunk.new(tampered[t["chunk_index"]].offset, hb(t["flipped_data_hex"]))
    assert_equal t["digest_hex"], Naalp::Streaming.commit_digest(tampered).unpack1("H*"),
                 "tampered stream digest == corpus tamper digest"
    err = assert_raises(Naalp::Streaming::StreamError) { Naalp::Streaming.verify_commit(commit, tampered) }
    assert_equal "StreamDigestMismatch", err.kind
  end

  # verify_checkpoint confirms a prefix without the end: contiguous from offset 0, totalling exactly
  # through_offset, digest matching. A non-contiguous prefix and a wrong through_offset both fail.
  def test_verify_checkpoint_prefix
    cp0 = C["checkpoints"][0]
    cp = Naalp::Streaming::StreamCheckpoint.new(stream_id, cp0["through_offset"], hb(cp0["digest_so_far_hex"]))
    prefix = corpus_chunks.select { |c| c.offset < cp0["through_offset"] }
    assert_nil Naalp::Streaming.verify_checkpoint(cp, prefix), "valid prefix confirms"

    # a non-contiguous prefix (drops the first chunk, so offset != 0) is rejected.
    noncontig = corpus_chunks.select { |c| c.offset >= cp0["through_offset"] }
    err = assert_raises(Naalp::Streaming::StreamError) { Naalp::Streaming.verify_checkpoint(cp, noncontig) }
    assert_equal "StreamDigestMismatch", err.kind
    # a checkpoint claiming the wrong through_offset over a valid prefix is rejected.
    bad = Naalp::Streaming::StreamCheckpoint.new(stream_id, cp0["through_offset"] + 1, hb(cp0["digest_so_far_hex"]))
    err2 = assert_raises(Naalp::Streaming::StreamError) { Naalp::Streaming.verify_checkpoint(bad, prefix) }
    assert_equal "StreamDigestMismatch", err2.kind
  end

  # A stream's effect is authorized BEFORE any chunk (R-10.3): an effect within the granted ceiling
  # proceeds; one exceeding it is refused fail-closed; an unrecognized effect is treated destructive.
  def test_open_stream_effect_authorization
    open = Naalp::Streaming::StreamOpen.new(stream_id, C["effect"], hb(C["approval_hex"]), C["substream"])
    assert_nil Naalp::Streaming.open_stream(open, Naalp::Policy::DESTRUCTIVE), "effect 1 under a destructive ceiling proceeds"
    err = assert_raises(Naalp::Streaming::StreamError) { Naalp::Streaming.open_stream(open, Naalp::Policy::READ_ONLY) }
    assert_equal "EffectNotAuthorized", err.kind
    # an unrecognized effect fails closed to destructive: it exceeds every non-destructive ceiling.
    wild = Naalp::Streaming::StreamOpen.new(stream_id, 99, nil, 0)
    err2 = assert_raises(Naalp::Streaming::StreamError) { Naalp::Streaming.open_stream(wild, Naalp::Policy::NON_IDEMPOTENT_WRITE) }
    assert_equal "EffectNotAuthorized", err2.kind
  end

  # NOT corpus-graded (no signature vector). Real deterministic ML-DSA-65 raw-signature round-trip in
  # isolation over each stream body; a flipped signature byte does not verify. Skips LOUDLY where
  # deterministic ML-DSA is unavailable.
  def test_full_signature_sign_verify_in_isolation
    begin
      pk = Naalp::COSE.mldsa_keygen("ML-DSA-65", SEED)
    rescue Exception => e
      skip "deterministic ML-DSA unavailable (needs OpenSSL >= 3.5): #{e.class}"
    end
    commit = Naalp::Streaming::StreamCommit.new(stream_id, hb(C["final_digest_hex"]))
    sig = Naalp::Streaming.sign_commit(commit, ALG, SEED)
    assert Naalp::Streaming.verify_commit_sig(commit, ALG, pk, sig), "the one end-commitment signature verifies"
    bad = sig.dup
    bad.setbyte(bad.bytesize - 1, bad.getbyte(bad.bytesize - 1) ^ 1)
    refute Naalp::Streaming.verify_commit_sig(commit, ALG, pk, bad), "a flipped signature byte does not verify"

    open = Naalp::Streaming::StreamOpen.new(stream_id, C["effect"], hb(C["approval_hex"]), C["substream"])
    osig = Naalp::Streaming.sign_open(open, ALG, SEED)
    assert Naalp::Streaming.verify_open_sig(open, ALG, pk, osig), "StreamOpen signature verifies"
  end
end
